File: src/g2si/g2si/g2si_node.py
import rclpy
from rclpy.node import Node
from sensor_msgs.msg import Imu
from geometry_msgs.msg import Vector3
import threading

class g2siNode(Node):
    def __init__(self):
        super().__init__('g2si_node')

        # Publisher to send corrected IMU data
        self.publisher_ = self.create_publisher(Imu, '/imu/data', 10)
        
        # Subscriber to receive raw IMU data
        self.subscription = self.create_subscription(Imu, '/imu/data_raw', self.listener_callback, 10)
        self.subscription

        # State variables
        self.data_received = False
        self.calibration_data = []  # Stores raw IMU data for calibration

        # Offsets for calibration (initialized to zero)
        self.accel_offsets = (0.0, 0.0, 0.0)
        self.gyro_offsets = (0.0, 0.0, 0.0)

        # Event to signal calibration completion
        self.calibration_event = threading.Event()

        # Timer to check if IMU data is received
        self.timer = self.create_timer(5.0, self.check_data_received)

        self.get_logger().info('Converting IMU linear acceleration from g\'s to m/s²')

    def listener_callback(self, msg):
        """
        Callback function to handle incoming IMU data.
        - Converts linear acceleration from g's to m/s².
        - Stores data for calibration if calibration is ongoing.
        - Publishes adjusted IMU data using the calculated offsets.
        """
        if not self.data_received:
            self.data_received = True
            self.get_logger().info('Found you! Received first IMU sample.')

        # During calibration, store raw IMU data for processing
        if not self.calibration_event.is_set():
            self.calibration_data.append((
                msg.linear_acceleration.x,
                msg.linear_acceleration.y,
                msg.linear_acceleration.z,
                msg.angular_velocity.x,
                msg.angular_velocity.y,
                msg.angular_velocity.z
            ))

        # Convert linear acceleration to m/s² using calibration offsets
        converted_acceleration = self.convert_g_to_ms2(msg.linear_acceleration)

        # Prepare and publish a new IMU message with adjusted values
        new_imu_msg = Imu()
        new_imu_msg.header = msg.header
        new_imu_msg.orientation = msg.orientation
        new_imu_msg.angular_velocity = msg.angular_velocity
        new_imu_msg.linear_acceleration = converted_acceleration

        self.publisher_.publish(new_imu_msg)

    def convert_g_to_ms2(self, accel_in_gs: Vector3) -> Vector3:
        """
        Converts acceleration from g's to m/s² and applies calibration offsets.
        :param accel_in_gs: Acceleration in g's (x, y, z components)
        :return: Adjusted acceleration in m/s²
        """
        conversion_factor = 9.81  # 1 g = 9.81 m/s²
        accel_in_ms2 = Vector3()

        # Apply calibration offsets and convert to m/s²
        accel_in_ms2.x = (accel_in_gs.x - self.accel_offsets[0]) * conversion_factor
        accel_in_ms2.y = (accel_in_gs.y - self.accel_offsets[1]) * conversion_factor
        accel_in_ms2.z = (accel_in_gs.z - self.accel_offsets[2]) * conversion_factor

        return accel_in_ms2

    def hl_calibrate_imu(self, num_samples=1000):
        """
        Performs IMU calibration by collecting data samples, calculating offsets, 
        and applying them for future adjustments.
        :param num_samples: Number of samples to collect for calibration.
        """
        self.get_logger().info("Starting IMU calibration... Keep sensor flat and steady! This is gonna take a hot second...")

        # Wait until enough samples are collected
        while len(self.calibration_data) < num_samples:
            rclpy.spin_once(self, timeout_sec=0.1)

        # Initialize accumulators
        ax = ay = az = gx = gy = gz = 0.0

        # Process collected data to calculate offsets
        for i in range(num_samples):
            try:
                ax += self.calibration_data[i][0]
                ay += self.calibration_data[i][1]
                az += self.calibration_data[i][2]
                gx += self.calibration_data[i][3]
                gy += self.calibration_data[i][4]
                gz += self.calibration_data[i][5]

            except Exception as e:
                self.get_logger().warning(f"Error during calibration data processing: {e}")
                continue

        # Calculate average offsets
        accel_offset_x = ax / num_samples
        accel_offset_y = ay / num_samples
        accel_offset_z = (az / num_samples) - 1.0 

        gyro_offset_x = gx / num_samples
        gyro_offset_y = gy / num_samples
        gyro_offset_z = gz / num_samples

        # Store calculated offsets
        self.accel_offsets = (accel_offset_x, accel_offset_y, accel_offset_z)
        self.gyro_offsets = (gyro_offset_x, gyro_offset_y, gyro_offset_z)

        self.get_logger().info("IMU Calibration complete!")
        self.get_logger().info(f"Accel Offsets: x={accel_offset_x}, y={accel_offset_y}, z={accel_offset_z}")
        self.get_logger().info(f"Gyro Offsets: x={gyro_offset_x}, y={gyro_offset_y}, z={gyro_offset_z}")

        # Signal calibration completion
        self.calibration_event.set()

    def check_data_received(self):
        """
        Checks if any IMU data has been received. If not, it logs a warning.
        """
        if not self.data_received:
            self.get_logger().warn('Hello? Is anyone out there? Still waiting on first IMU sample...')

def main(args=None):
    rclpy.init(args=args)
    node = g2siNode()

    # Optional: Perform calibration at startup
    node.hl_calibrate_imu()

    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/bird_navigation/src/g2si_node.cpp
#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/imu.hpp>
#include <geometry_msgs/msg/vector3.hpp>
#include <vector>
#include <deque>
#include <mutex>
#include <thread>
#include <condition_variable>

using namespace std::chrono_literals;

class G2siNode : public rclcpp::Node
{
public:
    G2siNode()
    : Node("g2si_node"),
      data_received_(false),
      calibration_complete_(false)
    {
        this->declare_parameter<bool>("use_cal", true);

        publisher_ = this->create_publisher<sensor_msgs::msg::Imu>("/imu/data", 10);

        subscription_ = this->create_subscription<sensor_msgs::msg::Imu>(
            "/imu/data_raw", 10,
            std::bind(&G2siNode::listenerCallback, this, std::placeholders::_1));

        timer_ = this->create_wall_timer(
            5s, std::bind(&G2siNode::checkDataReceived, this));

        RCLCPP_INFO(this->get_logger(), "Converting IMU linear acceleration from g's to m/s²");

        // ===== START calibration in background =====
        std::thread(&G2siNode::hlCalibrateIMU, this).detach();
    }

    void hlCalibrateIMU(size_t num_samples = 500)
    {
        bool use_cal;
        this->get_parameter("use_cal", use_cal);

        if (!use_cal)
        {
            RCLCPP_INFO(this->get_logger(), "Calibration disabled via parameter, skipping.");
            calibration_complete_ = true;
            return;
        }

        RCLCPP_INFO(this->get_logger(), "Starting IMU calibration... keep sensor flat & steady!");

        {
            std::unique_lock<std::mutex> lock(calibration_mutex_);
            if (!calibration_cv_.wait_for(lock, 120s, [this, num_samples]() {
                    return calibration_data_.size() >= num_samples;
                }))
            {
                RCLCPP_WARN(this->get_logger(), "Calibration timed out (got %zu of %zu samples).",
                            calibration_data_.size(), num_samples);
                calibration_complete_ = true;   // <— don’t block forever
                return;
            }
        }

        // compute offsets
        double ax=0, ay=0, az=0, gx=0, gy=0, gz=0;
        for (auto &d : calibration_data_)
        {
            ax += d[0]; ay += d[1]; az += d[2];
            gx += d[3]; gy += d[4]; gz += d[5];
        }
        size_t n = calibration_data_.size();
        accel_offsets_ = { ax/n, ay/n, (az/n)-1.0 };
        gyro_offsets_  = { gx/n, gy/n, gz/n };

        RCLCPP_INFO(this->get_logger(), "Calibration complete! Accel offsets (g): x=%.3f y=%.3f z=%.3f",
                    accel_offsets_[0], accel_offsets_[1], accel_offsets_[2]);
        RCLCPP_INFO(this->get_logger(), "Gyro offsets (rad/s): x=%.3f y=%.3f z=%.3f",
                    gyro_offsets_[0], gyro_offsets_[1], gyro_offsets_[2]);

        calibration_complete_ = true;
    }

private:
    void listenerCallback(const sensor_msgs::msg::Imu::SharedPtr msg)
    {
        if (!data_received_)
        {
            data_received_ = true;
            RCLCPP_INFO(this->get_logger(), "First IMU sample received.");
        }

        std::array<double,6> sample = {
            msg->linear_acceleration.x,
            msg->linear_acceleration.y,
            msg->linear_acceleration.z,
            msg->angular_velocity.x,
            msg->angular_velocity.y,
            msg->angular_velocity.z
        };

        // still calibrating? keep gathering and bail out
        if (!calibration_complete_)
        {
            {
                std::lock_guard<std::mutex> lock(calibration_mutex_);
                calibration_data_.push_back(sample);
                if (calibration_data_.size() % 50 == 0)
                    RCLCPP_INFO(this->get_logger(), "Collected %zu calibration samples",
                                calibration_data_.size());
            }
            calibration_cv_.notify_one();
            return;
        }

        // post‐calibration: sliding window
        moving_window_.push_back(sample);
        if (moving_window_.size() > max_window_size_)
            moving_window_.pop_front();

        // compute average
        std::array<double,6> avg = {0,0,0,0,0,0};
        for (auto &w : moving_window_)
            for (size_t i=0; i<6; ++i)
                avg[i] += w[i];
        for (size_t i=0; i<6; ++i)
            avg[i] /= moving_window_.size();

        geometry_msgs::msg::Vector3 acc_avg, gyr_avg;
        acc_avg.x = avg[0]; acc_avg.y = avg[1]; acc_avg.z = avg[2];
        gyr_avg.x = avg[3]; gyr_avg.y = avg[4]; gyr_avg.z = avg[5];

        // convert g→m/s²
        auto conv_acc = convertGToMs2(acc_avg);

        // publish filtered + calibrated IMU
        sensor_msgs::msg::Imu out{};
        out.header = msg->header;
        out.orientation = msg->orientation;
        out.angular_velocity = gyr_avg;
        out.linear_acceleration = conv_acc;
        publisher_->publish(out);
    }

    geometry_msgs::msg::Vector3 convertGToMs2(const geometry_msgs::msg::Vector3 &in)
    {
        constexpr double G = 9.81;
        geometry_msgs::msg::Vector3 out;
        out.x = (in.x - accel_offsets_[0]) * G;
        out.y = (in.y - accel_offsets_[1]) * G;
        out.z = (in.z - accel_offsets_[2]) * G;
        return out;
    }

    void checkDataReceived()
    {
        if (!data_received_)
            RCLCPP_WARN(this->get_logger(), "Waiting for IMU data...");
    }

    // ROS interfaces
    rclcpp::Publisher<sensor_msgs::msg::Imu>::SharedPtr publisher_;
    rclcpp::Subscription<sensor_msgs::msg::Imu>::SharedPtr subscription_;
    rclcpp::TimerBase::SharedPtr timer_;

    // state
    bool data_received_;
    bool calibration_complete_;
    std::vector<std::array<double,6>> calibration_data_;
    std::mutex calibration_mutex_;
    std::condition_variable calibration_cv_;

    std::array<double,3> accel_offsets_{0,0,0}, gyro_offsets_{0,0,0};

    // moving‐window
    std::deque<std::array<double,6>> moving_window_;
    const size_t max_window_size_ = 10;
};

int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<G2siNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}
